Add native call tracer

natebeauregard opened the pull request:
## Changes

Adds a native call tracer implementation to the nethermind client. The callTracer tracks all the call frames executed during a transaction, including depth 0. The result will be a nested list of call frames, resembling how EVM works. They form a tree with the top-level call at root and sub-calls as children of the higher levels.

Users can conditionally only select the main call or add emitted logs to the call stack as well depending on the tracerConfig that they pass in.

## Types of changes

#### What types of changes does your code introduce?

- [ ] Bugfix (a non-breaking change that fixes an issue)
- [x] New feature (a non-breaking change that adds functionality)
- [ ] Breaking change (a change that causes existing functionality not to work as expected)
- [ ] Optimization
- [x] Refactoring
- [ ] Documentation update
- [ ] Build-related changes
- [ ] Other: _Description_

## Testing

#### Requires testing

- [x] Yes
- [ ] No

#### If yes, did you write tests?

- [x] Yes
- [ ] No

## Documentation

#### Requires explanation in Release Notes

- [x] Yes
- [ ] No

The `ReportOperationLog` method and `IsTracingOpLevelLogs` flag was added to the `ITxTracer` interface to allow tracing logs emitted at the operation level.

natebeauregard commented on src/Nethermind/Nethermind.Evm/VirtualMachine.cs:
```diff
@@ -771,15 +771,33 @@ private CallResult ExecuteCall<TTracingInstructions>(EvmState vmState, ReadOnlyM
-            return _txTracer.IsTracingOpLevelStorage ?
-                ExecuteCode<TTracingInstructions, NotTracing, IsTracing>(vmState, ref stack, gasAvailable, spec) :
-                ExecuteCode<TTracingInstructions, NotTracing, NotTracing>(vmState, ref stack, gasAvailable, spec);
+            if (!_txTracer.IsTracingOpLevelStorage)
```
@benaadams Would you mind taking a look at these updates to add another shared generic parameter for tracing op level logs to the virtual machine since you initially implemented these in https://github.com/NethermindEth/nethermind/pull/5826? Since these if statements will now increase expontentially I worry that I'm setting a bad precedent here by adding another parameter.

benaadams replied:
Is just a single check inside `InstructionLog`? Don't think is work adding a another generic parameter for it (to cause branch elimination), might as well do the check in the method?

natebeauregard replied:
Yeah it's just a single check in `InstructionLog`. And sounds good, I'll remove these changes and just check the method directly instead

LukaszRozmej commented on src/Nethermind/Nethermind.Evm/Tracing/GethStyle/Custom/Native/Call/NativeCallTracerCallFrame.cs:
```diff
@@ -0,0 +1,37 @@
+
+    public Address? To { get; set; }
+
+    public byte[]? Input { get; set; }
+
+    public byte[]? Output { get; set; }
+
+    public string? Error { get; set; }
+
+    public string? RevertReason { get; set; }
+
+    public List<NativeCallTracerCallFrame> Calls { get; set; }
+
+    public List<NativeCallTracerLogEntry>? Logs { get; set; }
```
Potentially `Input`, `Output`, `Calls` and `Logs` could be replaced with `ArrayPoolList` if they would be then later correctly disposed. This would reduce memory allocations.

natebeauregard replied:
I attempted to spike out an approach for converting these fields to be of type `ArrayPoolList` instead, however I was having difficulty finding a way to only dispose of them after the trace object is serialized to JSON. I tried using the `Dispose` method from the `TxTracer` abstract class, however that would dispose of the lists before they were able to be serialized.

I looked around our other pre-existing traces (`GethLikeTxTrace`/`ParityLikeTxTrace`) to try to find an example of serializing `ArrayPoolList` objects, however it seems like they just use the built in list type (`[]`) or the `List` object type as well.

I agree that it would be nice to find a way to use `ArrayPoolList` for list type trace fields to reduce memory allocations, however it might make sense to move this into a separate issue due to the added complexity with disposing after serialization and also so we can optimize the existing opcode traces that store list type fields as well.

Here was my spiked approach for reference: [SPIKE](https://github.com/NethermindEth/nethermind/commit/59761b36b1449ff9498bbdb8db98a04488113174)

LukaszRozmej replied:
Our JSON RPC stack is built this way that if the serialized object is `IDisposable` then it's disposed will be called. as you can see `GethLikeTxTrace` is implementing disposable and allowing to pass custom internal one to get advanced logic. So you need to pass some custom disposable there that will dispose your `ArraPoolLists`

natebeauregard replied:
Ah ok that makes sense, thanks for the clarification. I'll make these updates to dispose of the `ArrayPoolList` fields at the call frame level instead of at the tracer level so that they are disposed after serialization.

LukaszRozmej reviewed (approved):
@benaadams we need your review on VirtualMachine stuff

natebeauregard commented on src/Nethermind/Nethermind.Evm/Tracing/CancellationTxTracer.cs:
```diff
@@ -213,6 +219,15 @@ public void ReportOperationRemainingGas(long gas)
+    public void ReportOperationLog(LogEntry log)
+    {
+        _token.ThrowIfCancellationRequested();
+        if (_innerTracer.IsTracingInstructions && _innerTracer.IsTracingOpLevelLogs)
```
Good point, I don't think that `IsTracingInstructions` is necessary here as well. I'll remove the `IsTracingInstructions` check

natebeauregard commented:
@LukaszRozmej I added a commit to update all of the list fields to use `ArrayPoolList` and then dispose of the array pools after the trace is serialized. Also, while conducting some final testing I noticed a bug where `SELFDESTRUCT` opcodes were being returned in a child call frame for the geth traces but not in nethermind so I added some additional logic to our call tracer to support that behavior.

Since some of these updates were non-trivial, especially in regards to the ArrayPoolList disposals, would you be able to look over the recent commits ([Use ArrayPoolList for list objects](https://github.com/NethermindEth/nethermind/pull/6945/commits/2b779d0bc83b6bba9277bc9aedc4da3ae3ff8343), [Fix selfdestruct bug](https://github.com/NethermindEth/nethermind/pull/6945/commits/c172fe4665066697fe0a954569f0c92154150e8b)) to see if you have any additional concerns prior to merging this?

LukaszRozmej commented on src/Nethermind/Nethermind.Evm/Tracing/GethStyle/Custom/Native/Call/NativeCallTracer.cs:
```diff
@@ -0,0 +1,262 @@
+        _firstCallFrame = _callStack[0];
+        result.CustomTracerResult = new GethLikeCustomTrace { Value = _firstCallFrame };
+
+        for (int i = 1; i < _callStack.Count; i++)
+        {
+            _callStack[i].Dispose();
+        }
```
I am a bit confused, we have a call stack that we dispose here and we dispose `NativeCallTracerCallFrame childCallFrame in Calls` in `NativeCallTracerCallFrame`, aren't we disposing same objects? What is the difference?

natebeauregard replied:
At this point in the code, `_callStack` should only contain one value, the top call frame. The first call frame is what's then passed into the trace and serialized into JSON. I added this loop as a protection against any potential memory leaks in case an unforseen error in the tracer causes the callStack to incorretly nest child elements and the elements are left in the top level.

However, this code wouldn't be necessary if I implement your suggestion above about making the call stack be an ArrayPoolList.

natebeauregard replied:
Paired and resolved synchronously

LukaszRozmej commented on src/Nethermind/Nethermind.Evm/Tracing/GethStyle/Custom/Native/Call/NativeCallTracerCallFrameConverter.cs:
```diff
@@ -0,0 +1,97 @@
+            }
+            else
+            {
+                JsonSerializer.Serialize(writer, value.Input.AsSpan().ToArray(), options);
```
why those calls are needed? ArrayPoolList should be easily serializable as collection as it implements `IList` interface. And it kind of beats the purpose of using it.

natebeauregard replied:
The ArrayPoolLists are serializable, however they'll most likely need a new converter to serialize their contents as hex. Currently the `ByteArrayConverter` only supports `byte[]` and `ReadOnlySpan<byte>` types for hex conversions. I'll look into making these changes to avoid needing to convert the ArrayPoolList

natebeauregard replied:
Paired and resolved synchronously

LukaszRozmej commented on src/Nethermind/Nethermind.Evm/Tracing/GethStyle/Custom/Native/Call/NativeCallTracer.cs:
```diff
@@ -0,0 +1,262 @@
+
+    private readonly long _gasLimit;
+    private readonly NativeCallTracerConfig _config;
+    private readonly List<NativeCallTracerCallFrame> _callStack = [];
```
Or if we would add `ParentFrame` into `NativeCallTracerCallFrame` we could remove `_callStack` alltogether and have it in a tree-structure kind of way?

natebeauregard replied:
Paired and resolved synchronously
